

---

#**Logistic Regression without using pandas**

---



In [15]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix, accuracy_score

In [2]:
# Load the breast cancer dataset
dataset = load_breast_cancer()

In [3]:
# Extract features and target variable
X = dataset.data
y = dataset.target

In [4]:
print(X)

[[1.799e+01 1.038e+01 1.228e+02 ... 2.654e-01 4.601e-01 1.189e-01]
 [2.057e+01 1.777e+01 1.329e+02 ... 1.860e-01 2.750e-01 8.902e-02]
 [1.969e+01 2.125e+01 1.300e+02 ... 2.430e-01 3.613e-01 8.758e-02]
 ...
 [1.660e+01 2.808e+01 1.083e+02 ... 1.418e-01 2.218e-01 7.820e-02]
 [2.060e+01 2.933e+01 1.401e+02 ... 2.650e-01 4.087e-01 1.240e-01]
 [7.760e+00 2.454e+01 4.792e+01 ... 0.000e+00 2.871e-01 7.039e-02]]


In [5]:
print(y)

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 1 0 0 0 0 0 0 0 0 1 0 1 1 1 1 1 0 0 1 0 0 1 1 1 1 0 1 0 0 1 1 1 1 0 1 0 0
 1 0 1 0 0 1 1 1 0 0 1 0 0 0 1 1 1 0 1 1 0 0 1 1 1 0 0 1 1 1 1 0 1 1 0 1 1
 1 1 1 1 1 1 0 0 0 1 0 0 1 1 1 0 0 1 0 1 0 0 1 0 0 1 1 0 1 1 0 1 1 1 1 0 1
 1 1 1 1 1 1 1 1 0 1 1 1 1 0 0 1 0 1 1 0 0 1 1 0 0 1 1 1 1 0 1 1 0 0 0 1 0
 1 0 1 1 1 0 1 1 0 0 1 0 0 0 0 1 0 0 0 1 0 1 0 1 1 0 1 0 0 0 0 1 1 0 0 1 1
 1 0 1 1 1 1 1 0 0 1 1 0 1 1 0 0 1 0 1 1 1 1 0 1 1 1 1 1 0 1 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 1 1 1 1 1 1 0 1 0 1 1 0 1 1 0 1 0 0 1 1 1 1 1 1 1 1 1 1 1 1
 1 0 1 1 0 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 0 1 0 1 1 1 1 0 0 0 1 1
 1 1 0 1 0 1 0 1 1 1 0 1 1 1 1 1 1 1 0 0 0 1 1 1 1 1 1 1 1 1 1 1 0 0 1 0 0
 0 1 0 0 1 1 1 1 1 0 1 1 1 1 1 0 1 1 1 0 1 1 0 0 1 1 1 1 1 1 0 1 1 1 1 1 1
 1 0 1 1 1 1 1 0 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 0 1 0 0 1 0 1 1 1 1 1 0 1 1
 0 1 0 1 1 0 1 0 1 1 1 1 1 1 1 1 0 0 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 0 1
 1 1 1 1 1 1 0 1 0 1 1 0 

In [6]:
# Split the dataset into training and testing sets
xtrain, xtest, ytrain, ytest = train_test_split(X, y, test_size=0.25, random_state=1)

In [7]:
# Initialize the Logistic Regression classifier
classifier = LogisticRegression()

In [8]:
# Train the model
classifier.fit(xtrain, ytrain)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [9]:
# Predict the target variable for the test set
predictions = classifier.predict(xtest)
print("Predictions:\n", predictions)

Predictions:
 [1 0 1 0 0 0 0 0 1 1 1 0 0 1 1 1 1 1 1 0 1 1 0 1 0 1 1 0 0 0 0 1 0 0 1 1 0
 1 0 1 1 1 1 1 1 0 1 1 1 0 0 0 1 1 1 1 1 0 1 1 1 0 1 0 1 1 1 0 1 1 1 1 1 0
 1 0 1 1 1 0 1 0 1 0 1 1 0 1 0 1 1 0 1 1 0 0 1 1 1 1 1 1 1 1 1 1 1 1 0 0 1
 1 0 1 0 0 1 1 1 1 1 0 0 1 1 0 0 0 0 0 1 1 1 0 1 0 0 1 1 0 0 0 1]


In [10]:
# Get the probabilities for each class
predictions_probability = classifier.predict_proba(xtest)
print("Predictions Probability (first 10):\n", predictions_probability[0:10])

Predictions Probability (first 10):
 [[4.59341099e-01 5.40658901e-01]
 [5.86836049e-01 4.13163951e-01]
 [2.75462204e-03 9.97245378e-01]
 [9.99598288e-01 4.01711773e-04]
 [8.43283792e-01 1.56716208e-01]
 [9.99171951e-01 8.28048931e-04]
 [9.99997958e-01 2.04223193e-06]
 [9.96876753e-01 3.12324746e-03]
 [1.94143338e-02 9.80585666e-01]
 [1.67342581e-02 9.83265742e-01]]


In [11]:
# Get the probability of class 0 (non-cancer)
predictions_probability_zero = classifier.predict_proba(xtest)[:, 0]
print("Probability of class 0 (first 10):\n", predictions_probability_zero[0:10])

Probability of class 0 (first 10):
 [0.4593411  0.58683605 0.00275462 0.99959829 0.84328379 0.99917195
 0.99999796 0.99687675 0.01941433 0.01673426]


In [12]:
# Get the probability of class 1 (cancer)
predictions_probability_one = classifier.predict_proba(xtest)[:, 1]
print("Probability of class 1 (first 10):\n", predictions_probability_one[0:10])

Probability of class 1 (first 10):
 [5.40658901e-01 4.13163951e-01 9.97245378e-01 4.01711773e-04
 1.56716208e-01 8.28048931e-04 2.04223193e-06 3.12324746e-03
 9.80585666e-01 9.83265742e-01]


In [13]:
# Calculate and print the confusion matrix
conf_matrix = confusion_matrix(ytest, predictions)
print("Confusion Matrix:\n", conf_matrix)

Confusion Matrix:
 [[51  4]
 [ 4 84]]


In [14]:
# Calculate and print the accuracy score
accuracy = accuracy_score(ytest, predictions)
print("Accuracy Score:\n", accuracy)

Accuracy Score:
 0.9440559440559441


---

#**Conclusion:**

The Logistic Regression model achieved an accuracy score of approximately **94.4%**, indicating strong performance in classifying breast cancer cases.

The confusion matrix shows that out of 139 test instances, the model correctly identified 51 non-cancerous cases and 84 cancerous cases, with only 4 false positives and 4 false negatives.

This suggests the model is effective in distinguishing between the two classes with high precision and recall.


---

